### PR TITLE
feat: add identity configuration module to the single application

### PR DESCRIPTION
**Pull request by Ben-Sheppard**
## Related issues
Closes https://github.com/camunda-cloud/identity/issues/2868, Closes https://github.com/camunda-cloud/identity/issues/2867

## Description
This PR introduces the Identity module into the single application.

I am aware that there are some tweaks required to the Operate module here as currently it adds the `auth` module by default when `Operate` is present and no other auth profile is found but I plan to address this in a follow up PR with some further Operate focused tweaks.

### CODE REVIEW

File: dist/src/main/resources/schema.sql
Comment (dlavrenuek) [inline dist/src/main/resources/schema.sql:10]: Apparently the schema file is now in `identity/common` (https://github.com/camunda/camunda/blob/93b38edf9bbaa79911f69df6d6904caa9a62bf2f/identity/common/src/main/resources/schema.sql), so it will be moved up?

Comment (Ben-Sheppard) [reply dist/src/main/resources/schema.sql:10]: Yea the schema file before this PR was in Identity-common but I moved it here because of it being relevant to all components (I'm happy to move it back though in my PR), either way I will rebase it based on Maryam's PR after she merges

File: dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java
Comment (dlavrenuek) [inline dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java:22]: When looking at the new architecture from a technical point of view, then `io.camunda.authentication` would belong to Camunda global/common. From ownership perspective it is fine having it here.

File: dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java
Comment (dlavrenuek) [inline dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java:30]: The authentication layer will also only be available with profile `identity` right? This should not be the case. So from this point of view it makes sense to have the auth classes being separated from the `identity` profile.

File: dist/src/main/resources/schema.sql
Comment (dlavrenuek) [reply dist/src/main/resources/schema.sql:10]: The current schema should then be removed, right? https://github.com/camunda/camunda/blob/main/identity/common/src/main/resources/schema.sql From the changes in this PR it looks like both files are still present and this one is not up to date after [the](https://github.com/camunda/camunda/pull/18773) was merged

Comment (dlavrenuek) [reply dist/src/main/resources/schema.sql:10]: ❓ Is the schema file automatically included via it's path `main/resources/schame.sql`?

**Review by dlavrenuek — CHANGES_REQUESTED** [review]

**Review by romansmirnov — COMMENTED** [review]
@Ben-Sheppard, thanks for opening the PR.

I have to admit, I don't know if I can approve the PR or not, because I lack some context of what we want to achieve with the profile `identity`. Or in other words, technically, it "works" but functionally I don't know what we want to achieve (and the linked issues do not provide much context therefore).

My naive assumption is that we want the following:
* When the profile `identity` is present, the application serves Identity's frontend (otherwise, not).
* For authentication, there exists separate profile(s) or configuration options that allow the user to configure what kind of authentication they want to use (disabling auth, oidc, basic auth, ...)

If my understanding is correct, then I think the component scanning must be different:
1. When profile `identity` is present, the component scanning configures the web application itself.
2. In terms of authentication, there would be a separate component scanning that scan the authentication modules accordingly.

However, could you please provide me with context?

File: dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java
Comment (Ben-Sheppard) [reply dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java:30]: That is correct, I added it this way because in my opinion having an auth layer without Identity didn't make sense, without Identity, the auth layer does not have a way to manage local users etc also, we currently have a hard dependency from `identity-backend` to the `auth-starter` because the Identity backend uses Spring security level classes `CamundaUserDetailsManager` for example. If we elect to only start Identity then the startup fails because of the missing bean that is in the `auth-starter` module.

I will work on trying to separate this though

**Comment by dlavrenuek** [comment]
As the new Identity will use Camunda API from the very beginning and all of our data will also be used in the API, all of the user/group etc services should always be included. Alternatively they could be hidden behind the `identity` profile for the start but should be included without conditions in final version.

Looking at Identity as a standalone application - in my opinion we would just serve our frontend and the Camunda rest api. It would be only necessary to server the endpoints that we actually use, but I did not see a concept within the unified rest api that only parts of it would be served

File: identity/backend/src/main/resources/schema.sql
Comment (romansmirnov) [inline identity/backend/src/main/resources/schema.sql:50]: How will it behave if multiple nodes try to create the schema simultaneously? Is there any locking applied?

File: identity/backend/src/main/resources/application.yaml
Comment (romansmirnov) [inline identity/backend/src/main/resources/application.yaml:1]: The settings at the "bottom" contradict with the single [application properties](https://github.com/camunda/camunda/blob/main/dist/src/main/resources/application.properties) settings, i.e., management port, context path, health endpoints, ... basically, everything under `management` and `server` must be removed.

```
management:
  server:
    port: 8082
  endpoints:
    web:
      exposure:
        include: health,prometheus
server:
  compression:
    enabled: true
  http2:
    enabled: true
  servlet:
    encoding:
      charset: UTF-8
      force: true
    context-path: ${IDENTITY_BASE_PATH:}
  max-http-request-header-size: ${SERVER_REQUEST_HEADER_MAX_SIZE:8KB}
```

File: identity/backend/src/main/resources/schema.sql
Comment (Ben-Sheppard) [reply identity/backend/src/main/resources/schema.sql:50]: We haven't specifically configured locking (the creation of the schema is left to the inner working of Spring), this is something we should consider but I'd rather address it as a follow up to unblock others with this PR

File: operate/qa/integration-tests/pom.xml
Comment (Ben-Sheppard) [inline operate/qa/integration-tests/pom.xml:53]: This is needed because with the dependency for some reason the tests were failing because classes were getting constructed vs mocked. For now I think this is fine though

File: authentication/pom.xml
Comment (romansmirnov) [inline authentication/pom.xml:49]:
```suggestion
```

File: authentication/pom.xml
Comment (romansmirnov) [inline authentication/pom.xml:54]:
```suggestion
```

File: authentication/pom.xml
Comment (romansmirnov) [inline authentication/pom.xml:59]:
```suggestion
```

File: authentication/pom.xml
Comment (romansmirnov) [inline authentication/pom.xml:19]: @Ben-Sheppard, a question for my understanding:
When going through the code of this module, I noticed that it uses the `JdbcUserDetailsManager` from Spring. To me, it feels like a Spring lock-in because they expect a specific DB schema, etc.

How flexible is it for extension? What happens if they change something that requires a different schema? A typical feature request from users is to configure table prefixes, would this still be possible with that approach?

(I know this is not related to this PR itself, but would be great, if you could guide me through it soon.)

File: dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java
Comment (romansmirnov) [inline dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java:36]: With `data-jpa` at the classpath, Spring started to scan all base packages defined in classes with the annotation `@EnableAutoConfiguration.` To avoid that Spring scan always all base packages, this setting ensures that it will only scan the passed identity package (and sub-packages). Could you please check if this is sufficient for Identity?

File: dist/src/main/resources/application.properties
Comment (romansmirnov) [inline dist/src/main/resources/application.properties:59]: @Ben-Sheppard, with adding additional spring dependencies, Spring does a lot of auto-configuration stuff. I excluded these once and imported them in `IdentityModuleConfiguration` so that they are only included when the identity module starts.

Could you please ensure (double-check) that Spring only auto-configures what is needed for Identity?

File: identity/backend/src/main/resources/application.yaml
Comment (romansmirnov) [inline identity/backend/src/main/resources/application.yaml:2]: @Ben-Sheppard, in the beginnings of the single application, we learned to have the `application` property files in a central place, otherwise, we run into issues that the application properties where loaded in a different order or some weren't applied at all, etc.

That's why, could you please move this configuration to the `application.properties` file in the `dist` module. That way, we have that in a central place, and do not run into issue in which order they are loaded.

File: identity/backend/src/main/resources/application.yaml
Comment (romansmirnov) [inline identity/backend/src/main/resources/application.yaml:5]: @Ben-Sheppard, this configuration seems to configure a default Jackson object mapper. Did you ensure that this configuration does not conflict with the other components? @houssain-barouni can help to spot the places where the default object mapper is used.

Comment (Ben-Sheppard) [reply identity/backend/src/main/resources/application.yaml:5]: This configuration was here from an early iteration, as we don't need to do anything different/custom it seems to me like I can just remove this configuration. I'll try that first and then reach out if any issues come up

File: dist/src/main/resources/schema.sql
Comment (Ben-Sheppard) [reply dist/src/main/resources/schema.sql]: Exactly, this is a known path to Spring so the file is automatically picked up

File: authentication/pom.xml
Comment (Ben-Sheppard) [reply authentication/pom.xml:19]: Hi @romansmirnov - thats correct, we elected to take the approach of aligning closely with Spring to reduce the amount of custom code required. As things currently are we can override the things we need to but fallback to the core for other things.

Having a different schema I accept at the moment is an issue however longer term I believe we should look for something like Flyway, this way we could manage the schemas and handle migrations when required. I hadn't heard of a configurable table prefix feature request before so this could well mean we need to rethink things.

My preference is that I'd like to keep it as it is for now to keep progressing but take the discussion to a separate context.

File: dist/src/main/resources/application.properties
Comment (Ben-Sheppard) [reply dist/src/main/resources/application.properties:58]: Hey @romansmirnov - I've looked here and I believe you have added all that is required for Identity, at least from my perspective, nothing else should be autoconfigured. I've also tested the app locally and it functions still.

File: dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java
Comment (Ben-Sheppard) [reply dist/src/main/java/io/camunda/identity/IdentityModuleConfiguration.java:36]: As I mentioned in the other comment @romansmirnov I've run the app locally and this configuration appears to be fine :)

**Review by dlavrenuek — APPROVED** [review]
Great job @Ben-Sheppard for the implementation and @romansmirnov for debugging ad fixing the Operate errors! 🚀

**Pull request merged**
